pangeo_forge_recipes/recipes/reference_hdf_zarr.py

```diff
@@ -0,0 +1,202 @@
+
+    @property
+    def finalize_target(self) -> Callable[[], None]:
+        """Final step to finish the recipe after data has been written.
+        Attribute that returns a callable function.
+        """
+        if isinstance(self.netcdf_url, str):
+            proto = fsspec.utils.get_protocol(self.netcdf_url)
+        else:
+            proto = fsspec.utils.get_protocol(self.netcdf_url[0])
+        return functools.partial(
+            _finalize,
+            work_dir=self._work_dir,
+            out_url=self.output_url,
+            out_so=self.output_storage_options,
+            remote_protocol=proto,
+            remote_options=self.netcdf_storage_options,
+            xr_open_kwargs=self.xarray_open_kwargs,
+            xr_concat_kwargs=self.xarray_concat_args,
+            template_count=self.template_count,
+        )
+
+
+def _one_chunk(of, work_dir):
+    with of as f:
+        fn = os.path.join(work_dir, os.path.basename(f.name + ".json"))
+        h5chunks = SingleHdf5ToZarr(f, _unstrip_protocol(f.name, f.fs), inline_threshold=300)
+        with open(fn, "w") as outf:
+            json.dump(h5chunks.translate(), outf)
+
+
+def _finalize(
+    work_dir,
+    out_url,
+    out_so,
+    remote_protocol,
+    remote_options,
+    xr_open_kwargs,
+    xr_concat_kwargs,
+    template_count,
+):
+    files = [os.path.join(work_dir, f) for f in os.listdir(work_dir)]
+    if len(files) == 1:
+        fn = files[0]
+    else:
+        mzz = MultiZarrToZarr(
+            files,
+            remote_protocol=remote_protocol,
+            remote_options=remote_options,
+            xarray_open_kwargs=xr_open_kwargs,
+            xarray_concat_args=xr_concat_kwargs,
+        )
+        fn = os.path.join(work_dir, "combined.json")
+        # mzz does not support directly writing to remote yet
+        mzz.translate(fn, template_count=template_count)
+    fs, _ = fsspec.core.url_to_fs(out_url, **out_so)
+    protocol = fs.protocol if isinstance(fs.protocol, str) else fs.protocol[0]
+    fs.put(fn, out_url)
+    fn2 = out_url + ".yaml"
+    spec = {
+        "sources": {
+            "data": {
+                "driver": "intake_xarray.xzarr.ZarrSource",
+                "description": "",  # could derive from data attrs or recipe
+                "args": {
+                    "urlpath": "reference://",
+                    "storage_options": {
+                        "fo": out_url,
+                        "target_protocol": protocol,
+                        "target_options": out_so,
+                        "remote_protocol": remote_protocol,
+                        "remote_options": remote_options,
+                        "skip_instance_cache": True,
+                    },
+                    "chunks": {},  # can optimize access here
+                    "consolidated": False,
+                },
+            }
+        }
+    }
+    with fs.open(fn2, "w") as f:
+        yaml.dump(spec, f, default_flow_style=False)
```

Comment:
Would be nice if we could use an`FSSpecTarget` here as the final location for the metadata to live.

Reply:
This is a remote URL. What does FSSpecTarget give me over `filesystem.open`?

Reply:
The bakeries assign targets to the recipes. Without accepting targets, your recipe will be unable to discover where to store its permanent and temporary data from the bakery.

Of course, we could resolve this by having the bakeries do different, custom things for each different recipes type. But I'd prefer not to go down that path. If your Recipe class has a `.target` and `.metadata_cache` attribute, the bakery can treat it the same as an `XarrayZarrRecipe`. Then you use these objects for your i/o. 

I'd welcome feedback on this design. So far we only have once recipe implementation, so many of the abstractions we have in mind are untested.  

> This is a remote URL. What does FSSpecTarget give me over `filesystem.open`?

One small but useful thing it does is hide the i/o behind a `root_path`, so that the recipe writer doesn't have to know the absolute URL of the final dataset. That is determined by the bakery.